### PR TITLE
Better schedule-importing and task-completion handling.

### PR DESCRIPTION
**Pull request by RonConescu**
Features:

1)  Only show a task in the "incomplete stuff" list on the first day it becomes "incomplete"

2)  If we get a new schedule for a task, and that task is a completed, one-time-only task, we ignore the new schedule

3)  Decomposed the import process into smaller, more maintainable components, moving it out of APCScheduler.m.  Now, we have:
    a)  ScheduleImporter:  top-level logic for importing schedules and tasks
    b)  ScheduleFilter and its subclasses:  analyze schedules during the import process
    c)  ScheduleTaskMap:  stores the results of CoreData queries needed by many of those Filters

### CODE REVIEW

**Comment by RonConescu** [comment]
@YuanZhu-apple — Just wanted to add your name to this, to make sure you saw it.  —ron

**Comment by YuanZhu-apple** [comment]
@RonConescu Sure, I am looking now.

File: APCAppCore/APCAppCore/Library/Categories/NSArray+APCHelper.m [hunk 1]
```diff
@@ -85,4 +69,16 @@ - (id) secondObject
...
+- (id) thirdObject
+{
+    id result = nil;
+
+    if (self.count >= 3)
+    {
+        result = self [2];
+    }
+
+    return result;
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/Library/Categories/NSArray+APCHelper.m:79]: `return (self.count >= 3) ? self [2] : nil;`

File: APCAppCore/APCAppCore/Library/Scheduler/APCScheduleImporter.m [hunk 2]
```diff
@@ -0,0 +1,1203 @@
...
+{
+    NSDateFormatter *formatter = [NSDateFormatter new];
+    formatter.locale = [NSLocale localeWithLocaleIdentifier: kAPCDateFormatLocaleEN_US_POSIX];
+
+    NSArray *legalFormats = @[@"H",
+                              @"HH",
+                              @"HH:mm",
+                              @"HH:mm:SS",
+                              @"HH:mm:SS.sss"
+                              ];
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/Library/Scheduler/APCScheduleImporter.m]: Can these `legalFormats` be shared between `serializing` and `deserializing`?

File: APCAppCore/APCAppCore/Library/Scheduler/ScheduleFilterClasses/APCCompletedOneTimeTaskFilter.m [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        for (APCTask *task in schedule.tasks)
+        {
+            APCScheduleTaskMapEntry *entry = [mapOfTheseTaskIdsToSavedTasksAndMostRecentSchedules entryForTaskId: task.taskID];
+
+            if (entry == nil)
+            {
+                /*
+                 Truly shouldn't happen.  We should only have been passed schedules
+                 and entries that go together.
+                 */
```
Comment (YuanZhu-apple) [inline APCAppCore/APCAppCore/Library/Scheduler/ScheduleFilterClasses/APCCompletedOneTimeTaskFilter.m]: Log something here, if behavior is not expected.

**Comment by RonConescu** [comment]
@YuanZhu-apple — Hi!  I just wanted to check in.  Are you waiting for me to fix the above, before you continue your analysis?  Or can you tell me:  once I fix the above, will this pull request be considered acceptable?  —ron

**Comment by YuanZhu-apple** [comment]
@RonConescu 
Once the fix is there, I accept this PR

**Comment by RonConescu** [comment]
Terrific.  Thanks.  In progress.

File: APCAppCore/APCAppCore/Library/Categories/NSArray+APCHelper.m (hunk 1 above)
Comment (RonConescu) [inline APCAppCore/APCAppCore/Library/Categories/NSArray+APCHelper.m:79]: Thanks.  I generalized the method to a more generic one.

**Comment by RonConescu** [comment]
@YuanZhu-apple - Good comments. Thanks. Here you go.

**Comment by YuanZhu-apple** [comment]
Good.

**Pull request merged**